### ISSUE TITLE
[bug] Requires ranges with version X.Y.Z.A sometimes can't be found in remote

### ISSUE THREAD

**Issue by albaltimore** [conan-io/conan#9805]
### Environment Details (include every applicable attribute)
  * Operating System+version: Windows Server 2019
  * Compiler+version: N/A
  * Conan version: 1.41.0
  * Python version: 3.8.2

### Steps to reproduce 

Some packages in our remote have versions of the form `X.Y.Z.A` (usually for third-party packages to which we've done small changes to so we know what was the base version and the `A` is versions of our changes).
However, sometimes it seems like Conan can't find the package when using ranges of the form `[>=X.Y.Z]` during the `conan lock create` command.

### Logs (Executed commands with output)

`conanfile.py` in the `test_conan` directory:
```python
from conans import ConanFile

class TestConan(ConanFile):
    name = "test-conan"

    def requirements(self):

        self.requires('aws-sdk-cpp/[>=1.6.53]@PORT/stable')
        self.requires('avro-cpp/[>=1.10.2]@PORT/stable')
```

Conan searches on the remote:
```
E:\alefebvr\test_conan>conan search aws-sdk-cpp -r bb-conan-buildall-stable
Existing package recipes:

aws-sdk-cpp/1.6.53.7@PORT/stable
```
```
E:\alefebvr\test_conan>conan search avro-cpp -r bb-conan-buildall-stable
Existing package recipes:

avro-cpp/1.9.1.9@PORT/stable
avro-cpp/1.9.1.10@PORT/stable
avro-cpp/1.10.2.0@PORT/stable
```

When running `conan lock create` command:
```
E:\alefebvr\test_conan>conan lock create ./conanfile.py --version 12345 --user PORT --channel stable --build=missing --lockfile-out=base_recipe_information_only.lock --base

ERROR: Version range '>=1.6.53' from requirement 'aws-sdk-cpp/[>=1.6.53]@PORT/stable' required by 'conanfile.py (test-conan/12345@PORT/stable)' could not be resolved in remote 'bb-conan-buildall-stable'
```

After commenting the line with `self.requires('aws-sdk-cpp/[>=1.6.53]@PORT/stable')` in the `conanfile.py`, rerunning the above command:
```
E:\alefebvr\test_conan>conan lock create ./conanfile.py --version 12345 --user PORT --channel stable --build=missing --lockfile-out=base_recipe_information_only.lock --base
[...]
Version ranges solved
    Version range '>=1.10.2' required by 'conanfile.py (test-conan/12345@PORT/stable)' resolved to 'avro-cpp/1.10.2.0@PORT/stable' in local cache
[...]

Requirements
    avro-cpp/1.10.2.0@PORT/stable from 'bb-conan-buildall-stable' - Cache
[...]
Packages
    avro-cpp/1.10.2.0@PORT/stable:460c0b2d584d7832de0ef56497dc51330405ecca - Cache
[...]

Generated lockfile: E:\alefebvr\test_conan\base_recipe_information_only.lock
```

**Comment by memsharded** [comment]
Hi @albaltimore 

At the moment this is not a bug. Version ranges resolve for the official semver specification, which contain 3 numbers, not 4. They cannot contain letters either. We are relying on a third party library that implements the semver standard, and this is quite difficult to change.

We are considering to extend to new versioning schemas for Conan 2.0, but that will require extra effort, because the third party library doesn't implement it, and we would need to do it ourselves. 

Labeling this as a feature request.

**Comment by albaltimore** [comment]
Thank you @memsharded for the quick answer! Any idea why sometimes it works and sometimes it doesn't?

**Comment by memsharded** [comment]
> Thank you @memsharded for the quick answer! Any idea why sometimes it works and sometimes it doesn't?

Yes, the 4th digit is basically ignored. There could be some randomness in the data structures ordering associated to order of creation, upload, or just luck that makes it work some times.

**Comment by michaelmaguire** [comment]
> At the moment this is not a bug. Version ranges resolve for the official semver specification, which contain 3 numbers, not 4. They cannot contain letters either. 

So no range comparisons for our beta or rc builds...

I love that semantic versioning doesn't follow semantic versioning:

https://semver.org/spec/v2.0.0-rc.2.html

**Comment by michaelmaguire** [comment]
> They cannot contain letters either. 

@albaltimore pointed out https://semver.org/#spec-item-9 to me:

> A pre-release version MAY be denoted by appending a hyphen and a series of dot separated identifiers immediately following the patch version. Identifiers MUST comprise only ASCII alphanumerics and hyphens [0-9A-Za-z-].

So I guess letters can be present?  It seems like it ought not to be too hard (80/20 rule?) to do some smarter things with Conan versions in range comparisons so as not to assume they adhere to semver.  E.g. with `1.2.3.4` < `1.2.3.5`, `1.2.3-rc1` < `1.2.3-rc2`.  Also, OpenSSL users could sure be happy if we could do `1.1.1p` < `1.1.1r`.

> We are relying on a third party library that implements the semver standard, and this is quite difficult to change.

I was going to ask which library, but I found:

https://github.com/conan-io/conan/issues/5366 "Semantic Versioning - 4 digits [question]"
https://github.com/conan-io/conan/pull/6941 "lets try to update node-semver"

I see https://pypi.org/project/node-semver/ now shows release:  

0.8.0
 handle 4-digit version correctly (#35) 
 
 I see https://github.com/podhmo/python-node-semver/pull/35 has:
 
 uilianries mentioned this pull request on 5 May 2020
lets try to update node-semver conan-io/conan#6941


Did it not work out?

Note: I see now that even the earliest versions of the [`version` attribute](https://github.com/conan-io/docs/commit/f01091e5daae17bf6daf8ad1b25d1df6ead0574e) and ["version ranges" documentation](https://github.com/conan-io/docs/blob/6d02f56ba126441e1473d2390f5ee2f2fb24ddb4/mastering/version_ranges.rst) were very clear right from the beginning that ranges are only going to work if you follow semver -- my oversight.

**Comment by memsharded** [comment]
We have taken ownership of version-ranges and implemented our own solution, covering more than 3 canonical major.minor.patch. This is solved and released in 2.0-alpha, please try it when possible and report if necessary.
